carma_wm/src/CARMAWorldModel.cpp

```diff
@@ -1061,4 +1081,126 @@ void CARMAWorldModel::setConfigSpeedLimit(double config_lim)
+                           ", and signal_group_id: " << (int)current_movement_state.signal_group);
+        continue;
+      }
+      lanelet::CarmaTrafficLightPtr curr_light = curr_light_list[0];
+      
+      // reset states if the intersection's geometry changed 
+      if (curr_light->revision_ != curr_intersection.revision)
+      {
+        ROS_INFO_STREAM("Received a new intersection geometry. intersection_id: " << (int)curr_intersection.id.id << 
+                           ", and signal_group_id: " << (int)current_movement_state.signal_group);
+        traffic_light_states_[curr_intersection.id.id].clear();
+      }
+
+      // currently expecting 1 lane, so 0th index is used only when setting states
+      if (current_movement_state.movement_event_list.empty())
+      {
+        ROS_INFO_STREAM("movement_event_list is empty . intersection_id: " << (int)curr_intersection.id.id << 
+                           ", and signal_group_id: " << (int)current_movement_state.signal_group);
+        continue;
+      }
+
+      traffic_light_states_[curr_intersection.id.id][current_movement_state.signal_group].push_back(std::make_pair(ros::Time(current_movement_state.movement_event_list[0].timing.min_end_time), 
+                              static_cast<lanelet::CarmaTrafficLightState>(current_movement_state.movement_event_list[0].event_state.movement_phase_state)));
+      
+      if (traffic_light_states_[curr_intersection.id.id][current_movement_state.signal_group].size() >= 2 
+          && traffic_light_states_[curr_intersection.id.id][current_movement_state.signal_group].front().second == 
+             traffic_light_states_[curr_intersection.id.id][current_movement_state.signal_group].back().second)
+          {
+            curr_light->setStates(traffic_light_states_[curr_intersection.id.id][current_movement_state.signal_group], curr_intersection.revision);
+          }
+      else if (curr_light->recorded_time_stamps.empty()) // if it was never initialized, do its best to plan with the current state until the future state is also received.
+      {
+        std::vector<std::pair<ros::Time, lanelet::CarmaTrafficLightState>> default_state;
+        // green 20sec, yellow 3sec, red 20sec, back to green 20sec etc...
+        default_state.push_back(std::make_pair<ros::Time, lanelet::CarmaTrafficLightState>(ros::Time(0), lanelet::CarmaTrafficLightState::PERMISSIVE_MOVEMENT_ALLOWED));
+        default_state.push_back(std::make_pair<ros::Time, lanelet::CarmaTrafficLightState>(ros::Time(3), lanelet::CarmaTrafficLightState::PRE_MOVEMENT));
+        default_state.push_back(std::make_pair<ros::Time, lanelet::CarmaTrafficLightState>(ros::Time(23), lanelet::CarmaTrafficLightState::STOP_AND_REMAIN));
+        default_state.push_back(std::make_pair<ros::Time, lanelet::CarmaTrafficLightState>(ros::Time(43), lanelet::CarmaTrafficLightState::PERMISSIVE_MOVEMENT_ALLOWED));
+        curr_light->setStates(default_state, curr_light->revision_);
+      }
```

Comment:
There is a missing case here. 
When you have the full cycle you are setting it. When you have no cycle you are using defaults, but we also need the case where you set the partial cycle. 
For example if you default is (in length of phase) 30s g, 5s y, 10s r
Then when you receive a Spat with green phase of 20s duration you should update the cycle to have durations of 
20s g, 5s y, 10s r

When the yellow is then received as a duration of 8s you would update the known phases to 
20s g, 8s y, 10s r

Ditto for red